specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+
+DataSet.is_empty
+    This attribute will be true if the DataSet is empty (has no results), or false if at least one result has been added to the DataSet.
+    It is equivalent to testing if the length is zero.
+
+DataSet.is_completed
+    This attribute will be true if the DataSet is completed or false if it is in progress.
+
+DataSet.get_data(*params, start=, end=)
+    Returns the values stored in the DataSet for the specified parameters.
+    The values are returned as a list of parallel NumPy arrays, one array per parameter.
+    The data type of each array is based on the data type provided when the DataSet was created.
+    If a parameter is optional and no value was provided for one or more results, the corresponding array entries will be the “null” value for the data type: zero for integers, NaN for floats, “” for strings, None for objects.
+    The parameter list may contain a mix of string parameter names, QCoDeS Parameter objects, and ParamSpec objects.
+    If provided, the start and end parameters select a range of results by result count (index). 
+    Start defaults to 0, and end defaults to the current length.
+    If the range is empty -- that is, if the end is less than or equal to the start – then a list of empty arrays is returned.
+
+DataSet.get_parameters()
+    Returns a list of ParamSpec objects that describe the parameters stored in this DataSet.
+
+DataSet.get_metadata(tag=)
+    Returns metadata for this DataSet.
+    If a tag string is provided, only metadata stored under that tag is returned.
+    Otherwise, all metadata is returned.
+
+DataSet.subscribe(callback, state=)
+    Subscribes the provided callback function to result additions to the DataSet.
+    Every time one or more results are added to the DataSet, the callback is called.
+    It is passed the DataSet itself, the length of the DataSet before the triggering addition, the length after the addition, and the state object provided when subscribing.
+    If no state object was provided, then the callback gets passed None as the fourth parameter.
+    When the DataSet is completed, the callback gets called with the length of the DataSet as both the before and after lengths.
+    This method returns an opaque subscription identifier.
+
+DataSet.unsubscribe(subid)
+    Removes the indicated subscription.
+    The subid must be the same object that was returned from a DataSet.subscribe call.
+
+Storage
+-------
+
+DataSet.read_from(location, formatter=)
+    Reads a DataSet from persistent store.
+    Location may be a string file system path, a string URL, or some other string that is meaningful to the formatter specified.
+    Formatter is a QCoDeS Formatter object that specifies how data is read and written. 
+    If not provided, the default formatter is used. 
+    The default formatter is currently GNUPlotFormat().
+    This is a static method in the DataSet class.
+    It returns a new DataSet object.
+
+DataSet.read_updates()
```

Comment:
I am a bit worried about the implications of having a read_updates() command. 

Such a feature is only useful if the data in the underlying location is expected to change, and as such suggests the possibility of having multiple copies of the same dataset open. This may be exactly what we want (I can think of numerous applications where it is useful to have the DataSet open in another proces) but we should then also ensure we properly manage any potential for conflicting changes to the data.